src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnusedLocalVariableCheck.java

```diff
@@ -0,0 +1,449 @@
+        TokenTypes.POST_INC,
+        TokenTypes.POST_DEC,
+        TokenTypes.INC,
+        TokenTypes.DEC,
+    };
+
+    /**
+     * An array of scope tokens.
+     */
+    private static final int[] SCOPES = {
+        TokenTypes.SLIST,
+        TokenTypes.LITERAL_IF,
+        TokenTypes.LITERAL_FOR,
+        TokenTypes.LITERAL_WHILE,
+    };
+
+    /**
+     * An array of unacceptable children of ast of type {@link TokenTypes#DOT}.
+     */
+    private static final int[] UNACCEPTABLE_CHILD_OF_DOT = {
+        TokenTypes.DOT,
+        TokenTypes.METHOD_CALL,
+        TokenTypes.LITERAL_NEW,
+        TokenTypes.LITERAL_SUPER,
+        TokenTypes.LITERAL_CLASS,
+        TokenTypes.LITERAL_THIS,
+    };
+
+    /**
+     * An array of unacceptable parent of ast of type {@link TokenTypes#IDENT}.
+     */
+    private static final int[] UNACCEPTABLE_PARENT_OF_IDENT = {
+        TokenTypes.VARIABLE_DEF,
+        TokenTypes.DOT,
+        TokenTypes.LITERAL_NEW,
+        TokenTypes.PATTERN_VARIABLE_DEF,
+        TokenTypes.METHOD_CALL,
+        TokenTypes.TYPE,
+    };
+
+    /**
+     * Keeps VariableDesc objects inside stack of declared variables.
+     */
+    private Deque<VariableDesc> variables;
+
+    @Override
+    public int[] getDefaultTokens() {
+        return new int[] {
+            TokenTypes.DOT,
+            TokenTypes.VARIABLE_DEF,
+            TokenTypes.IDENT,
+            TokenTypes.SLIST,
+            TokenTypes.LITERAL_IF,
+            TokenTypes.LITERAL_FOR,
+            TokenTypes.LITERAL_WHILE,
+        };
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getDefaultTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return getDefaultTokens();
+    }
+
+    @Override
+    public void beginTree(DetailAST root) {
+        variables = new ArrayDeque<>();
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        final int type = ast.getType();
+        if (type == TokenTypes.DOT) {
+            visitDotToken(ast);
+        }
+        else if (type == TokenTypes.VARIABLE_DEF) {
+            visitVariableDefToken(ast);
+        }
+        else if (type == TokenTypes.IDENT) {
+            visitIdentToken(ast);
+        }
+    }
+
+    @Override
+    public void leaveToken(DetailAST ast) {
+        if (TokenUtil.isOfType(ast, SCOPES)) {
+            while (!variables.isEmpty() && variables.peek().getScope() == ast) {
+                final VariableDesc variableDesc = variables.pop();
+                if (!variableDesc.isUsed()) {
+                    final DetailAST typeAst = variableDesc.getTypeAst();
+                    log(typeAst, MSG_UNUSED_LOCAL_VARIABLE, variableDesc.name);
+                }
+            }
+        }
+    }
+
+    /**
+     * Visit ast of type {@link TokenTypes#DOT}.
+     *
+     * @param dotAst dotAst
+     */
+    private void visitDotToken(DetailAST dotAst) {
+        if (shouldCheckIdentifierAst(dotAst)) {
```

Comment:
`shouldCheckIdentifierAst(dotAst)` sounds confusing, please rename argument and method `visitDotToken` since there can be not only dot

Reply:
1. I renamed it to `shouldCheckCurrentAndNestedIdentAst(DetailAST ast)`, renaming the argument of this method to something like `identOrDotAst` didn't feel right to me as every time it was used there was an ambiguity whether it is `ident` or `dot` ast.

2. It didn't rename the method `visitDotToken` since we are visiting dot token first and then checking it surrounding tokens using some methods, so it feels right to me.

If renaming in the first point doesn't work I can split the method into two separate methods, one for ident ast and the other for dot ast.

Reply:
you can always make arg name like `ast`, this is common way

Reply:
Right now arg name for `shouldCheckCurrentAndNestedIdentAst(..)` is `ast`

Reply:
I mean here it is dotAst, but it is not correct (there can be ident). So just make it more abstract to avoid confusion.
In general, it is better to avoid specifying exact tokens in method/arg name unless it is 100% the only token that can be there. Otherwise, when arg is named like `dotAst` and we check its type to be IDENT, it is a bit confusing.

Reply:
Done, I refactored into two different methods to make more sense.